staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1426,35 +1434,120 @@ func (c *cacheWatcher) process(ctx context.Context, resourceVersion uint64) {
 	}
 }
 
+// ready is a condition variable that represent readiness and
+// allow callers to wait on its status, returning an error with the reason.
+// It allows to use context cancellation and to be stopped,
+// on both cases it stops waiting raising the corresponding error.
+
 type ready struct {
-	ok bool
-	c  *sync.Cond
+	mu      sync.Mutex
+	readyCh chan struct{}
+	stopCh  chan struct{}
 }
 
+// newReady creates a ready condition variable that is not stopped and not ready
 func newReady() *ready {
-	return &ready{c: sync.NewCond(&sync.RWMutex{})}
+	return &ready{
+		readyCh: make(chan struct{}), // is ready if it does not exist
+		stopCh:  make(chan struct{}), // is stopped if it does not exist
```

Comment:
I'm not strongly against it, but I'm wondering if we can' achieve what we want by setting to not-ready instead?

Reply:
if not ready we block on the "readyCh" on wait, so we need to unblock on stop without deleting the readyCh, that will mean that is ready.